Fail if inputs contain nested nulls in min/max::toIntermediate

duanmeng opened the pull request:
`toIntermediate`  is an adaptive optimization to abandon partial
aggregation that has no meaningful cardinality reduction. For min/max
aggregates `toIntermediate` act as follows,

- Raw data in raw data out (sometimes with dictionary encoding)
- It happens in step ofStep::kPartial
- Its result is the input of another step Step::kIntermediate\Step::kFinal,
  which would not check nested nulls of the input, causing no null checks for
  the rest of the raw input.

This is a follow-up to #6723. `NonNumericMinMaxAggregateBase::toIntermediate`
also needs to check complex type inputs and throw if there are nested nulls.

mbasmanova reviewed (commented):
@duanmeng Thank you for the quick fix. Looks great.

Any chance you could add a test?

I think it can be as simple as modifying existing tests to pass multiple vectors to testAggregations. This will trigger a code path in testAggregations that tests abandon-partial-aggregation optimization.

```
VELOX_ASSERT_THROW(
      testAggregations({data, data, data}, {}, {"min(c0)", "max(c0)"}, {expected}), ...)
```

duanmeng commented:
> @duanmeng Thank you for the quick fix. Looks great.
> 
> Any chance you could add a test?
> 
> I think it can be as simple as modifying existing tests to pass multiple vectors to testAggregations. This will trigger a code path in testAggregations that tests abandon-partial-aggregation optimization.
> 
> ```
> VELOX_ASSERT_THROW(
>       testAggregations({data, data, data}, {}, {"min(c0)", "max(c0)"}, {expected}), ...)
> ```

@mbasmanova Got it, thanks. I've added a test to trigger `toIntermediate` according to this suggestion.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/MinMaxTest.cpp:
```diff
@@ -431,6 +431,40 @@ TEST_F(MinMaxTest, rowCheckNull) {
+
+  for (const auto& expr : {"min(c0)", "max(c0)"}) {
+    VELOX_ASSERT_THROW(
+        testToIntermediate({batch, batchWithNull}, {"c1"}, {expr}),
```
Looks like testAggregations cannot be used to effectively test failure scenarios. Let's not introduce testToIntermediate, but rather add testFailingAggregations and make sure that this method verifies that all equivalent plans fail with the specified error message.

duanmeng replied:
Got it, done.

mbasmanova commented on velox/functions/lib/aggregates/tests/AggregationTestBase.h:
```diff
@@ -206,6 +206,15 @@ class AggregationTestBase : public exec::test::OperatorTestBase {
         });
   }
 
+  /// Use abandon-partial-aggregation optimization to trigger the toIntermediate
```
Shouldn't this method be more generic and verify that all logically-equivalent plans fail with the specified error message?

duanmeng replied:
I forgot to update the comments. Fixed.

mbasmanova commented on velox/functions/lib/aggregates/tests/AggregationTestBase.h:
```diff
@@ -206,6 +206,15 @@ class AggregationTestBase : public exec::test::OperatorTestBase {
+      const std::vector<std::string>& groupingKeys,
+      const std::vector<std::string>& aggregates,
+      const std::unordered_map<std::string, std::string>& config = {},
+      const std::string& expectedMessage = "");
```
Let's move this error message so that it doesn't have a default value. There is no point in calling this method without specifying the error message.

One option is to move expectedMessage to be the first argument. Another to move it just before config.

duanmeng replied:
Fixed.

mbasmanova commented on velox/functions/lib/aggregates/tests/AggregationTestBase.cpp:
```diff
@@ -999,4 +1000,96 @@ VectorPtr AggregationTestBase::testStreaming(
+        .intermediateAggregation()
+        .finalAggregation();
+    AssertQueryBuilder queryBuilder(builder.planNode());
+    queryBuilder
```
nit: this code repeats, right? Consider extracting a helper function to avoid copy-paste

```
enableAbandonPartialAggregation(queryBuilder);
```

duanmeng replied:
Got it, done.

duanmeng commented:
@mbasmanova Hi Masha, I've added the testFailingAggregations method and refactored the two test cases `arrayCheckNull` and `rowCheckNull`. They would verify that all equivalent plans fail with the specified error message including `toIntermediate` scenario.

mbasmanova reviewed (commented):
Thanks. Looks good % a few nits.

mbasmanova reviewed (approved):
Thanks.
